src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,187 @@
 using System.Linq;
+using System.Runtime;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
 using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.Optimizations;
+using Microsoft.Quantum.QsCompiler.SymbolManagement;
 using Microsoft.Quantum.QsCompiler.SyntaxTokens;
 using Microsoft.Quantum.QsCompiler.SyntaxTree;
 using Microsoft.Quantum.QsCompiler.Transformations.Core;
 
 namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
 {
+    using Concretion = HashSet<(QsTypeParameter, ResolvedType)>;
+
+    internal struct Request
+    {
+        public QsQualifiedName originalName;
+        public Concretion typeResolutions;
+        public QsQualifiedName concreteName;
+    }
+
+    internal struct Response
+    {
+        public QsQualifiedName originalName;
+        public Concretion typeResolutions;
+        public QsCallable concreteCallable;
+    }
+
+    #region ResolveGenerics
+
     public class ResolveGenericsSyntax :
-        SyntaxTreeTransformation<ScopeTransformation<ResolveGenericsExpression>>
+        SyntaxTreeTransformation<NoScopeTransformations>
     {
-        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces)
+        ImmutableDictionary<NonNullable<string>, IEnumerable<QsCallable>> NamespaceCallables;
+
+        public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces, QsQualifiedName entryPointName)
         {
             if (namespaces == null || namespaces.Contains(null)) throw new ArgumentNullException(nameof(namespaces));
 
-            // Get list of generics
-            List<QsQualifiedName> generics = new List<QsQualifiedName>();
-            var Callables = namespaces.GlobalCallableResolutions();
-            foreach (var callable in Callables)
+            var globals = namespaces.GlobalCallableResolutions();
+
+            var entryPoints = globals
+                .Where(call => // TODO: get list of entry points
+                    call.Key.Namespace.Value == entryPointName.Namespace.Value &&
+                    call.Key.Name.Value == entryPointName.Name.Value)
+                .Select(call => new Request
+                {
+                    originalName = call.Key,
+                    typeResolutions = null,
+                    concreteName = call.Key
+                });
+
+            var requests = new Stack<Request>(entryPoints);
+            var responses = new List<Response>();
+
+            while(requests.Count() > 0)
             {
-                if (callable.Value.Signature.TypeParameters.Any())
+                Request currentRequest = requests.Pop();
+
+                // If there is a call to an unknown callable, throw exception
+                if (!globals.ContainsKey(currentRequest.originalName)) throw new ArgumentNullException(); // TODO: need to throw a more valid exception
```

Comment:
How about addressing the todo? 
And I'd suggest to go with a globals.TryGetValue to save the duplicate effort to first check if the key exists and then actually get the value further below.

Reply:
The problem is I don't know what kind of exception to throw here.